lean/bitvec.lean

```diff
@@ -0,0 +1,325 @@
+
+-- 2s complement negation
+protected def neg {n:ℕ} (x : bitvec n) : bitvec n :=
+  ⟨if x.val = 0 then 0 else 2^n - x.val,
+   begin
+     by_cases (x.val = 0),
+     { simp [h], apply pos_pow_of_pos, dec_trivial_tac },
+     { simp [h],
+       have pos : 0 < 2^n - x.val, { apply nat.sub_pos_of_lt x.property },
+       have x_val_pos: 0 < x.val, { apply nat.pos_of_ne_zero h },
+       apply sub_lt_of_pos_le x.val (2^n) x_val_pos,
+       apply le_of_lt x.property,
+     }
+    end⟩
+
+section conversion
+  -- Operations for converting to/from bitvectors
+  variable {α : Type}
+
+  protected def to_nat {n : nat} (v : bitvec n) : nat := v.val
+
+  protected def to_int {n:ℕ} (x: bitvec n) : int :=
+    match msb x with
+    | ff := int.of_nat x.val
+    | tt := int.neg_of_nat (bitvec.neg x).val
+  end
+
+  protected def of_nat (n : ℕ) (x:ℕ) : bitvec n := ⟨ x % 2^n, by simp ⟩
+
+  protected def of_int : Π (n : ℕ), int → bitvec (succ n)
+  | n (int.of_nat m)          := bitvec.of_nat (succ n) m
+  | n (int.neg_succ_of_nat m) := bitvec.neg (bitvec.of_nat (succ n) m)
+
+  theorem of_nat_to_nat {n : ℕ} (x : bitvec n)
+  : bitvec.of_nat n (bitvec.to_nat x) = x :=
+    begin
+      cases x,
+      simp [bitvec.to_nat, bitvec.of_nat],
+      apply mod_eq_of_lt x_property,
+    end
+
+  theorem to_nat_of_nat {k n : ℕ}
+  : bitvec.to_nat (bitvec.of_nat k n) = n % 2^k :=
+    begin
+      simp [bitvec.of_nat, bitvec.to_nat]
+    end
+
+end conversion
+
+section bitwise
+
+  -- bitwise negation
+  def not {w:ℕ} (x: bitvec w) : bitvec w := ⟨2^w - x.val - 1,
+    begin
+      have xval_pos : 0 < x.val + 1,
+      { apply (succ_pos x.val) },
+      apply (sub_lt _ xval_pos),
+      apply pos_pow_of_pos,
+      apply (succ_pos 1)
+    end⟩
+
+  -- logical bitwise and
+  def and {w:ℕ} (x y : bitvec w) : bitvec w := bitvec.of_nat w (nat.land x.val y.val)
+  -- logical bitwise or
+  def or  {w:ℕ} (x y : bitvec w) : bitvec w := bitvec.of_nat w (nat.lor  x.val y.val)
+  -- logical bitwise xor
+  def xor {w:ℕ} (x y : bitvec w) : bitvec w := bitvec.of_nat w (nat.lxor x.val y.val)
+
+  infix `.&&.`:70 := and
+  infix `.||.`:65 := or
+
+end bitwise
+
+section arith
+  -- Arithmetic operations on bitvectors
+  variable {n : ℕ}
+
+  -- Add with carry (no overflow)
+  def adc (x y : bitvec n) (c : bool) : bitvec n × bool :=
+    let c₁ := if c then 1 else 0,
+        r  := x.val + y.val + c₁ in
+    ⟨ bitvec.of_nat n r, r ≥ 2^n ⟩
+
+  protected def add (x y : bitvec n) : bitvec n := (adc x y ff).1
+
+  -- Subtract with borrow
+  def sbb (x y : bitvec n) (b : bool) : bool × bitvec n :=
+    let b₁ : bitvec n := if b then 1 else 0,
+        r  := match bitvec.adc x (bitvec.neg y) ff with
+              | (z, b₂) := bitvec.adc z (bitvec.neg b₁) ff
+              end
+    in ⟨ if b then y.val + 1 > x.val else y.val > x.val , r.1 ⟩
+
+  -- Usual arithmetic subtraction
+  protected def sub (x y : bitvec n) : bitvec n := (sbb x y ff).2
```

Comment:
I think I'd write `sbb` as

```
 def sbb (x y : bitvec n) (b : bool) : bool × bitvec n :=
   ⟨ x.to_nat < y.to_nat + (if b then 1 else 0)
   , of_int (x.to_int - (y.to_int + (if b then 1 else 0))) n
   ⟩
```

I'd also defined `add` and `sub` directly rather than calling `adc` and `sbb`.  That made since when we had to go through vectors anyways, but less so now.